File: Ejercicios-de-Repaso/ej3.hs
{-
Implementar la funcion dineroEnStock :: [(String, Int)] ->[(String, Float)] ->Float
problema dineroEnStock (stock: seq⟨String × Z⟩, precios: seq⟨String × R⟩ ) : R {
    requiere: {No existen dos nombres de productos (primeras componentes) iguales en stock}
    requiere: {No existen dos nombres de productos (primeras componentes) iguales en precios}
    requiere: {Todas las cantidades (segundas componentes) de stock son mayores a cero}
    requiere: {Todos los precios (segundas componentes) de precios son mayores a cero}
    requiere: {Todo producto de stock aparece en la lista de precios}
    asegura: {res es igual a la suma de los precios de todos los productos que estan en stock multiplicado por la cantidad de cada producto que hay en stock}
}
Para resolver este ejercicio pueden utilizar la funci´on del Preludio de Haskell fromIntegral que dado un valor de tipo
Int devuelve su equivalente de tipo Float.

-}
stock :: [(String, Int)]
stock = [("Agua", 3), ("Coca", 3), ("Papel Higienico", 2), ("Crema", 4), ("Doritos", 1)]

prices :: [(String, Float)]
prices = [("Agua", 2500.50), ("Coca", 3000.0), ("Papel Higienico", 1500.0), ("Crema", 5000.0), ("Doritos", 1000.0)]

dineroEnStock :: [(String, Int)] -> [(String, Float)] -> Float
dineroEnStock [] _ = 0
dineroEnStock ((prodStock, stock) : otrosProdsStock) prodsPrices = dineroEnStockDeUnSoloProducto + dineroEnStockDeLosOtrosProductos
  where
    dineroEnStockDeUnSoloProducto = dineroEnStockDeUnProducto (prodStock, stock) prodsPrices
    dineroEnStockDeLosOtrosProductos = dineroEnStock otrosProdsStock prodsPrices

dineroEnStockDeUnProducto :: (String, Int) -> [(String, Float)] -> Float
dineroEnStockDeUnProducto _ [] = 0 -- Por especificacion no deberia suceder
dineroEnStockDeUnProducto (prodStock, stock) ((prodPrice, price) : otrosProdsPrices)
  | prodStock == prodPrice = fromIntegral stock * price
  | otherwise = dineroEnStockDeUnProducto (prodStock, stock) otrosProdsPrices

File: Python/P7/ej3.py
"""
Implementar una función para conocer el estado de aprobación de una materia a partir de las notas obtenidas por un/a alumno/a cumpliendo con la siguiente especificación:

    problema resultado_materia (in notas: seq⟨Z⟩) : Z {
        requiere: { |notas| > 0 }
        requiere: { Para todo i ∈ Z si 0 ≤ i < |notas| → 0 ≤ notas[i] ≤ 10) }
        asegura: { res = 1 ↔ todos los elementos de notas son mayores o iguales a 4 y el promedio es mayor o igual a 7 }
        asegura: { res = 2 ↔ todos los elementos de notas son mayores o iguales a 4 y el promedio está entre 4 (inclusive) y 7 }
        asegura: { res = 3 ↔ alguno de los elementos de notas es menor a 4 o el promedio es menor a 4 }
    }
"""
def resultado_materia(notas:list[int]) -> int:
    promedio:float = 0
    notas_desaprobadas:int = 0
    for nota in notas:
        promedio += nota
        if nota < 4: notas_desaprobadas += 1

    promedio /= len(notas)
    res:int = 2
    if notas_desaprobadas == 0 and promedio >=7: res = 1
    if notas_desaprobadas != 0 or promedio < 4: res = 3

    return res
